### PR TITLE
LS-458 io-connectors

### PR DESCRIPTION
**Pull request by mendonk**

### CODE REVIEW

File: modules/ROOT/pages/io-connectors.adoc [hunk 1]
```diff
@@ -0,0 +1,55 @@
+= Luna Streaming IO Connectors
+
+When you have Luna Streaming installed and running, add IO connectors to your deployment to interact with external systems, such as Apache Cassandra, Aerospike, and many others.
```
Comment (nicoloboschi) [inline modules/ROOT/pages/io-connectors.adoc:3]: Aerospike is not included in Luna Streaming. See https://github.com/datastax/release-notes/blob/master/Luna_Streaming_2.8_Release_Notes.md#builtin-connectors

File: modules/ROOT/pages/io-connectors.adoc [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+== Source Connectors
+
+Source connectors pull messages _from_ external topics and persist the messages _to_ Pulsar topics.
+
+Explore source connector configurations and code examples https://pulsar.apache.org/docs/en/io-connectors/#source-connector[here].
```
Comment (nicoloboschi) [inline modules/ROOT/pages/io-connectors.adoc:9]: broken link ?

File: modules/ROOT/pages/io-connectors.adoc [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+== Sink Connectors
+
+Sink connectors are used to pull messages _from_ Pulsar topics and persist the messages _to_ external topics.
```
Comment (nicoloboschi) [inline modules/ROOT/pages/io-connectors.adoc:13]:
```suggestion
Sink connectors are used to pull messages _from_ Pulsar topics and persist the messages _to_ external systems.
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+Explore sink connector configurations and code examples https://pulsar.apache.org/docs/en/io-connectors/#sink-connector[here].
+
+== Processing Guarantees
```
Comment (nicoloboschi) [inline modules/ROOT/pages/io-connectors.adoc:17]: Why don't we just point out to the Pulsar official docs ?

**Review by nicoloboschi — CHANGES_REQUESTED** [review]

File: modules/ROOT/pages/io-connectors.adoc [hunk 5]
```diff
@@ -0,0 +1,110 @@
+= Luna Streaming IO Connectors
+
+When you have Luna Streaming installed and running, add IO connectors to your deployment to interact with external systems, such as Apache Cassandra, Elastic Search, and many others.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:3]:
```suggestion
When you have Luna Streaming installed and running, add IO connectors to your deployment to interact with external systems, such as Apache Cassandra®, Elasticsearch, and others.
```

@mendonk Make sure you use ® for first use of Apache Cassandra on each page and format Elasticsearch properly each time (no space, lowercase S).

File: modules/ROOT/pages/io-connectors.adoc [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+== Source Connectors
+
+Source connectors pull messages _from_ external topics and persist the messages _to_ Pulsar topics.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:7]:
```suggestion
Source connectors pull messages from external topics and persist the messages to Apache Pulsar™ topics.
```

Emphasis isn't required. Include "Apache" and ™ for first reference of Pulsar on a page.

File: modules/ROOT/pages/io-connectors.adoc [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+Explore source connector configurations and code examples https://pulsar.apache.org/docs/en/io-connectors/#source-connector[here].
+
+== Sink Connectors
+
+Sink connectors are used to pull messages _from_ Pulsar topics and persist the messages _to_ external systems.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:13]:
```suggestion
Sink connectors are used to pull messages from Pulsar topics and persist the messages to external systems.
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 8]
```diff
@@ -0,0 +1,110 @@
+= Luna Streaming IO Connectors
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:1]:
```suggestion
= Luna Streaming IO connectors
```

Capitalize first word and proper nouns only. See below also.

File: modules/ROOT/pages/io-connectors.adoc [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+When you have Luna Streaming installed and running, add IO connectors to your deployment to interact with external systems, such as Apache Cassandra, Elastic Search, and many others.
+
+== Source Connectors
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:5]:
```suggestion
== Source connectors
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+Explore source connector configurations and code examples https://pulsar.apache.org/docs/en/io-connectors/#source-connector[here].
+
+== Sink Connectors
+
+Sink connectors are used to pull messages _from_ Pulsar topics and persist the messages _to_ external systems.
+
+Explore sink connector configurations and code examples https://pulsar.apache.org/docs/en/io-connectors/#sink-connector[here].
+
+== Processing Guarantees
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:17]:
```suggestion
== Processing guarantees
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+Processing guarantees are used to handle errors when writing messages to Pulsar topics.
+
+* *ATLEAST_ONCE*: Each message sent to the source can be processed more than once.
+* *ATMOST_ONCE*: The message sent to the source is processed at most once. Therefore, there is a chance that the message is not processed.
+* *EFFECTIVELY_ONCE*: Each message sent to the source will have one output associated with it.
+
+Learn more about processing guarantees https://pulsar.apache.org/docs/en/io-overview/#processing-guarantee[here].
+
+== Kafka Connectors
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:27]:
```suggestion
== Kafka connectors
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+Datastax offers https://github.com/datastax/pulsar-3rdparty-connector[third-party connector support] for creating Pulsar connectors on the base of existing Kafka connectors. This allows developers to quickly move connectors from their Apache Kafka infrastructure into Apache Pulsar.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:29]:
```suggestion
DataStax offers https://github.com/datastax/pulsar-3rdparty-connector[third-party connector support] for creating Pulsar connectors on the base of existing Kafka connectors. This allows developers to quickly move connectors from their Apache Kafka® infrastructure into Pulsar.
```

DataStax has a capital D and S. Use ® for first use of Kafka on page.

File: modules/ROOT/pages/io-connectors.adoc [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+**Prerequisites:**
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:31]:
```suggestion
== Prerequisites
```

Prereq title should be at same level of the procedure headings.

File: modules/ROOT/pages/io-connectors.adoc [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+* Ensure you have **JDK 11+** and **Maven 3.8** installed.
+
+* Clone the connector's repo and `run mvn clean install` from the root.
+
+* The connector's `.nar` files can be found at `pulsar-connectors/<connector name>>/target/pulsar-3rdparty-pulsar-connectors-<connector name>-0.1.0-SNAPSHOT.nar`.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:37]:
```suggestion
* JDK 11+ installed.
* Maven 3.8 installed.
* Clone the connector's repo. Run `mvn clean install` from the root.

The connector's `.nar` files can be found at `pulsar-connectors/<connector name>/target/pulsar-3rdparty-pulsar-connectors-<connector name>-0.1.0-SNAPSHOT.nar`.
```

I think the command is just `mvn clean install`. Please confirm

File: modules/ROOT/pages/io-connectors.adoc [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+=== Configuring connectors
+
+Follow the documentation https://pulsar.apache.org/docs/en/io-use/[here] to initialize your connector. Then, modify its `config.yaml` file as below to create an adapter for the connector. For further information on shading and troubleshooting, see the https://github.com/datastax/pulsar-3rdparty-connector[GitHub readme].
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:41]: @mendonk Let's give them enough basic steps they can do it here. Then we can point to the pulsar docs for more info.

File: modules/ROOT/pages/io-connectors.adoc [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+**Prerequisites:**
+
+* Ensure you have **JDK 11+** and **Maven 3.8** installed.
+
+* Clone the connector's repo and `run mvn clean install` from the root.
+
+* The connector's `.nar` files can be found at `pulsar-connectors/<connector name>>/target/pulsar-3rdparty-pulsar-connectors-<connector name>-0.1.0-SNAPSHOT.nar`.
+
+=== Configuring connectors
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:39]:
```suggestion
== Configuring connectors
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+
+Follow the documentation https://pulsar.apache.org/docs/en/io-use/[here] to initialize your connector. Then, modify its `config.yaml` file as below to create an adapter for the connector. For further information on shading and troubleshooting, see the https://github.com/datastax/pulsar-3rdparty-connector[GitHub readme].
+
+==== Sink Connectors
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:43]:
```suggestion
=== Sink connector `config.yaml` example
```

Is this section title more accurate?

File: modules/ROOT/pages/io-connectors.adoc [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    # The following properties passed directly to Kafka Connect Sink and defined by it or by
+    # https://github.com/apache/kafka/blob/2.7/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java
+    name: "test-sink"
+    connector.class: "com.third.party.CoolSinkConnector"
+    tasks.max: "1"
+    topics: "my-topic"
+    ...
+----
+
+==== Source Connectors
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:72]:
```suggestion
=== Source connectors `config.yaml` example
```

**Review by jgillenwater — CHANGES_REQUESTED** [review]

File: modules/ROOT/pages/io-connectors.adoc [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+
+Processing guarantees are used to handle errors when writing messages to Pulsar topics.
+
+* *ATLEAST_ONCE*: Each message sent to the source can be processed more than once.
+* *ATMOST_ONCE*: The message sent to the source is processed at most once. Therefore, there is a chance that the message is not processed.
+* *EFFECTIVELY_ONCE*: Each message sent to the source will have one output associated with it.
+
+For more about processing guarantees, see https://pulsar.apache.org/docs/en/io-overview/#processing-guarantee[processing guarantees^].
+
+== Adding third party connectors
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:27]:
```suggestion
== Adding third-party connectors
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+
+Datastax offers https://github.com/datastax/pulsar-3rdparty-connector[third-party connector support^] for creating Pulsar connectors on the base of existing Kafka connectors. This allows developers to quickly move connectors from their Apache Kafka infrastructure into Apache Pulsar.
+
+=== Prerequisites
+
+. https://www.oracle.com/java/technologies/downloads/[JDK 11+^] installed.
+
+. https://maven.apache.org/download.cgi?Preferred=ftp://ftp.osuosl.org/pub/apache/[Maven 3.8^] installed.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:35]:
```suggestion
. https://www.oracle.com/java/technologies/downloads/[JDK 11+^] installed.
. https://maven.apache.org/download.cgi?Preferred=ftp://ftp.osuosl.org/pub/apache/[Maven 3.8^] installed.
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+
+== Building connectors
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:37]:
```suggestion
=== Building connectors
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+
+. Add the new module into `pulsar-connectors/pom.xml`.
+
+. Update the `pulsar-connectors/<connector name>/README.md`.
+
+. Build  with `mvn clean install`. 
+
+. Run `mvn dependency:tree -verbose` to review how Maven auto-resolved potential dependency conflicts. Fix as needed.
+
+For further information on shading and troubleshooting, see the https://github.com/datastax/pulsar-3rdparty-connector[Third party connectors GitHub readme^].
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:142]:
```suggestion
For more information on shading and troubleshooting, see the https://github.com/datastax/pulsar-3rdparty-connector[Third-party connectors GitHub readme^].
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+Source connectors pull messages from external topics and persist the messages to Apache Pulsar™ topics.
+
+For more source connector configurations and code examples, see https://pulsar.apache.org/docs/en/io-connectors/#source-connector[Pulsar built-in connectors^].
+
+== Sink connectors
+
+Sink connectors pull messages from Pulsar topics and persist the messages to external systems.
+
+For more sink connector configurations and code examples, see https://pulsar.apache.org/docs/en/io-connectors/#sink-connector[Pulsar built-in connectors^].
+
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:16]:
```suggestion
* *Source connectors:* Source connectors pull messages from external topics and persist the messages to Apache Pulsar™ topics. For more, see https://pulsar.apache.org/docs/en/io-connectors/#source-connector[Pulsar built-in connectors^].

* *Sink connectors:* Sink connectors pull messages from Pulsar topics and persist the messages to external systems. For more, see https://pulsar.apache.org/docs/en/io-connectors/#sink-connector[Pulsar built-in connectors^].
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+== Processing guarantees
+
+Processing guarantees are used to handle errors when writing messages to Pulsar topics.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:19]:
```suggestion
Processing guarantees handle errors when writing messages to Pulsar topics:
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+
+. Clone the connector's repo. 
+
+. Run `mvn clean install` from the root.
+
+. The connector's `.nar` files can be found at `pulsar-connectors/<connector name>>/target/pulsar-3rdparty-pulsar-connectors-<connector name>-0.1.0-SNAPSHOT.nar`.
+
+. Copy the connector's `.nar` file to the `connectors` directory in the Pulsar director. 
+
+. Modify the connector's `config.yaml` file to match the connector YAML files below. This creates an adapter for the connector.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:47]:
```suggestion
. Clone the connector's repo. 
. Run `mvn clean install` from the root.
. Copy the connector's `.nar` file to the `connectors` directory in the Pulsar director.
+
[NOTE]
====
The connector's `.nar` files can be found at `pulsar-connectors/<connector_name>/target/pulsar-3rdparty-pulsar-connectors-<connector_name>-0.1.0-SNAPSHOT.nar`.
====

. Modify the connector's `config.yaml` file to match the connector YAML files below. 
+
This creates an adapter for the connector. 
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+
+. Copy the connector's `.nar` file to the `connectors` directory in the Pulsar director. 
+
+. Modify the connector's `config.yaml` file to match the connector YAML files below. This creates an adapter for the connector. 
+
+For more, see https://pulsar.apache.org/docs/en/standalone/#install-builtin-connectors-optional[Install built-in connectors^].
+
+=== Sink connector `config.yaml` example
+
+[SOURCE, shell]
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:53]:
```suggestion
[source, shell]
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+    name: "test-sink"
+    connector.class: "com.third.party.CoolSinkConnector"
+    tasks.max: "1"
+    topics: "my-topic"
+    ...
+----
+
+=== Source connector `config.yaml` example
+
+[SOURCE, shell]
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:83]:
```suggestion
[source, shell]
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 28]
```diff
@@ -0,0 +1,150 @@
...
+  key.converter: "org.apache.kafka.connect.json.JsonConverter"
+  ...
+----
+
+== Troubleshooting
+
+Some common problems encountered while using KCA to create a new connector:
+
+. Ensure the connector's license allows its use and redistribution. For more, see the https://www.apache.org/legal/resolved.html[Apache third-party license policy^].
+. Maven dependency conflict of transitive dependencies in build time, or dependency conflict in runtime caused by third-party dependencies packaged with the connector. To resolve third-party dependencies, see xref:io-connectors.adoc#shading[Shading third-party connectors^].
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:119]:
```suggestion
These are common solutions for problems encountered while using KCA to create a new connector:

. Ensure the connector's license allows its use and redistribution. For more, see the https://www.apache.org/legal/resolved.html[Apache third-party license policy^].
. Resolve third-party dependencies if there is a Maven dependency conflict of transitive dependencies in build time or dependency conflict in runtime caused by third-party dependencies packaged with the connector. To resolve third-party dependencies, see xref:io-connectors.adoc#shading[Shading third-party connectors^].
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 29]
```diff
@@ -0,0 +1,150 @@
...
+
+[#shading]
+=== Shading third-party connectors
+
+If your connector includes third-party dependencies, you may need to rename some classes. This is called "shading".
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:124]:
```suggestion
If your connector includes third-party dependencies, you may need to rename some classes. This is called "shading."
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+
+. Copy `shaded-dependencies/template-shaded/` to `shaded-dependencies/<connector name>` and add the new module into `shaded-dependencies/pom.xml`.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:126]:
```suggestion
. Copy `shaded-dependencies/template-shaded/` to `shaded-dependencies/<connector_name>`. 
. Add the new module into `shaded-dependencies/pom.xml`.
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 31]
```diff
@@ -0,0 +1,150 @@
...
+
+. Copy `pulsar-connectors/template/` to `pulsar-connectors/<connector name>/`.
+
+. Add the new module into `pulsar-connectors/pom.xml`.
+
+. Update the `pulsar-connectors/<connector name>/README.md`.
+
+. Build  with `mvn clean install`. 
+
+. Run `mvn dependency:tree -verbose` to review how Maven auto-resolved potential dependency conflicts. Fix as needed.
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:140]:
```suggestion
. Ensure third-party dependencies are renamed as specified in `shaded-dependencies/<connector name>/pom.xml` 
. Build with `mvn clean install`.
. Copy `pulsar-connectors/template/` to `pulsar-connectors/<connector_name>/`.
. Add the new module into `pulsar-connectors/pom.xml`.
. Update the `pulsar-connectors/<connector name>/README.md`.
. Build with `mvn clean install`. 
. Run `mvn dependency:tree -verbose` to review how Maven auto-resolved potential dependency conflicts. Fix as needed.
```

File: modules/ROOT/pages/io-connectors.adoc [hunk 32]
```diff
@@ -0,0 +1,150 @@
...
+
+For further information on shading and troubleshooting, see the https://github.com/datastax/pulsar-3rdparty-connector[Third party connectors GitHub readme^].
+
+=== Next
+
+Learn more about Pulsar IO connectors https://pulsar.apache.org/docs/en/io-overview/[here^].
+
+
+
```
Comment (jgillenwater) [inline modules/ROOT/pages/io-connectors.adoc:149]:
```suggestion
Learn more about Pulsar IO connectors https://pulsar.apache.org/docs/en/io-overview/[here^].
```

**Review by jgillenwater — APPROVED** [review]

**Pull request merged**